File: src/reasoners/tot.py
from __future__ import annotations

import heapq
import re
from typing import Optional

from .base import Reasoner, ReasoningTrace


class ToTReasoner(Reasoner):
    def solve(self, query: str, initial_solution: Optional[str] = None) -> ReasoningTrace:
        beam = self.config.get("beam_width", 6)
        max_depth = self.config.get("max_depth", 10)
        branch = self.config.get("branch_factor", 3)

        if initial_solution:
            frontier = [(initial_solution, self._score(query, initial_solution))]
        else:
            frontier = [("", 0.0)]

        best_reasoning = ""
        best_answer = ""
        best_score = float("-inf")
        evaluated = 0

        for _depth in range(max_depth):
            new = []
            for reasoning, score_prev in frontier:
                if "final answer" in reasoning.lower():
                    if score_prev > best_score:
                        best_score = score_prev
                        best_reasoning = reasoning
                        best_answer = self._extract_answer(reasoning)
                    continue

                prompt = f"""{query}

Reasoning so far:
{reasoning or "(empty)"}

Continue reasoning. If done, end with 'Final answer: ...'.
"""
                for _ in range(branch):
                    cont = self._chat(
                        [{"role": "user", "content": prompt}],
                        temperature=0.9,
                        max_tokens=512,
                    )
                    full = (reasoning + "\n" + cont).strip()
                    score = self._score(query, full)
                    evaluated += 1

                    if "final answer" in full.lower() and score > best_score:
                        best_score = score
                        best_reasoning = full
                        best_answer = self._extract_answer(full)

                    new.append((full, score))

            if not new:
                break

            frontier = heapq.nlargest(beam, new, key=lambda item: item[1])

        if best_score == float("-inf") and frontier:
            best_reasoning, best_score = max(frontier, key=lambda item: item[1])
            best_answer = self._extract_answer(best_reasoning)

        return ReasoningTrace(
            best_reasoning,
            best_answer,
            {"method": "tot", "states": evaluated},
        )

    def _score(self, query: str, reasoning: str) -> float:
        prompt = f"""
[Problem]
{query}

[Partial solution]
{reasoning}

Rate promise (0-1). Output only a number.
"""
        raw = self._chat(
            [{"role": "user", "content": prompt}],
            temperature=0.0,
            max_tokens=16,
        )
        match = re.search(r"[0-1](?:\.\d+)?", raw.strip())
        if match:
            try:
                return float(match.group(0))
            except Exception:
                pass
        length = min(len(reasoning) / 1000, 1.0)
        bonus = 0.2 if "final answer" in reasoning.lower() else 0.0
        return length + bonus


__all__ = ["ToTReasoner"]

File: tui/packages/agent-core/src/reasoners/tot.ts
// ── ToT: Tree of Thoughts ──────────────────────────────────────────────────────
// Adapted from Python src/reasoners/tot.py.
//
// Beam search over reasoning paths, scoring each candidate.

import { BaseReasoner, type ReasoningTrace } from "./base.js";

interface ToTConfig {
	beamWidth?: number;
	maxDepth?: number;
	branchFactor?: number;
	temperature?: number;
	maxTokens?: number;
	[key: string]: unknown;
}

export class ToTReasoner extends BaseReasoner {
	config: ToTConfig;

	constructor(
		llm: import("../core/backend.ts").LLMBackend,
		config: ToTConfig = {},
	) {
		super(llm, config);
		this.config = config;
	}

	async solve(
		query: string,
		initialSolution?: string | undefined | null,
	): Promise<ReasoningTrace> {
		const beam = this.config.beamWidth ?? 6;
		const maxDepth = this.config.maxDepth ?? 10;
		const branch = this.config.branchFactor ?? 3;

		type State = [reasoning: string, score: number];
		let frontier: State[] = initialSolution
			? [[initialSolution, await this._score(query, initialSolution)]]
			: [["", 0.0]];

		let bestReasoning = "";
		let bestAnswer = "";
		let bestScore = -Infinity;
		let evaluated = 0;

		for (let depth = 0; depth < maxDepth; depth++) {
			const next: State[] = [];
			for (const [reasoning, scorePrev] of frontier) {
				if (/final answer/i.test(reasoning)) {
					if (scorePrev > bestScore) {
						bestScore = scorePrev;
						bestReasoning = reasoning;
						bestAnswer = BaseReasoner._extractAnswer(reasoning);
					}
					continue;
				}
				const prompt = `${query}\n\nReasoning so far:\n${reasoning || "(empty)"}\n\nContinue reasoning. If done, end with 'Final answer: ...'.`;
				for (let b = 0; b < branch; b++) {
					const cont = await this._chat([{ role: "user", content: prompt }], {
						temperature: 0.9,
						maxTokens: 512,
					});
					const full = `${reasoning}\n${cont}`.trim();
					const score = await this._score(query, full);
					evaluated++;
					if (/final answer/i.test(full) && score > bestScore) {
						bestScore = score;
						bestReasoning = full;
						bestAnswer = BaseReasoner._extractAnswer(full);
					}
					next.push([full, score]);
				}
			}
			if (next.length === 0) break;
			frontier = next.sort((a, b) => b[1] - a[1]).slice(0, beam);
		}

		if (bestScore === -Infinity && frontier.length > 0) {
			const [best] = frontier.sort((a, b) => b[1] - a[1]);
			bestReasoning = best[0];
			bestAnswer = BaseReasoner._extractAnswer(best[0]);
		}

		return {
			reasoning: bestReasoning,
			answer: bestAnswer,
			metadata: { method: "tot", states: evaluated },
		};
	}

	private async _score(query: string, reasoning: string): Promise<number> {
		const prompt = `[Problem]\n${query}\n\n[Partial solution]\n${reasoning}\n\nRate promise (0-1). Output only a number.`;
		const raw = await this._chat([{ role: "user", content: prompt }], {
			temperature: 0.0,
			maxTokens: 16,
		});
		const match = raw.trim().match(/[0-1](?:\.\d+)?/);
		if (match) {
			const v = parseFloat(match[0]);
			if (!Number.isNaN(v)) return v;
		}
		const length = Math.min(reasoning.length / 1000, 1.0);
		const bonus = /final answer/i.test(reasoning) ? 0.2 : 0.0;
		return length + bonus;
	}
}
